[scipy.linalg](https://docs.scipy.org/doc/scipy/reference/linalg.html)

Example 4.13 (p.270)

## Compute det A and B if $
A=\left [
\begin{matrix}
2 & -3 & -1\\
0 &  5 &  3\\
-4& -6 & 2\\
\end{matrix}
\right ]
$, $
B=\left [
\begin{matrix}
0 & 2 & -4 & 5\\
3 & 0 & -3 & 6\\
2 & 4 & 5 & 7\\
5 & -1 & -3 & 1\\
\end{matrix}
\right ]
$ ?  


In [2]:
import numpy as np

In [3]:
A=np.array([[2,3,-1],
           [0,5,3],
           [-4,-6,2]])

In [4]:
from scipy import linalg


In [5]:
linalg.det(A)

-1.9984014443252818e-15

In [6]:
B=np.array([[0,2,-4,5],
           [3,0,-3,6],
           [2,4,5,7],
           [5,-1,-3,1]])
linalg.det(B)

585.0000000000001

Example 4.18 (p.292)

## Find the eigenvalues and the corresponding eigenspaces of $ A = \left [
\begin{matrix}
0 & 1 & 0\\
0 &  0 & 1\\
2& -5 & 4\\
\end{matrix}
\right ]
$

In [14]:
A=np.array([[0,1,0],
           [0,0,1],
           [2,-5,4]])
eig=np.linalg.eig(A)
print(eig)
eigvalues,eigvectors=eig[0],eig[1]

(array([1.00000001, 0.99999999, 2.        ]), array([[-0.57735026,  0.57735028, -0.21821789],
       [-0.57735027,  0.57735027, -0.43643578],
       [-0.57735028,  0.57735026, -0.87287156]]))


Example 4.19 (p.294)

## Find the eigenvalues and the corresponding eigenspaces of $ A = \left [
\begin{matrix}
-1 & 0 & 1\\
3 &  0 & -3\\
1 & 0 & -1\\
\end{matrix}
\right ]
$ 

In [15]:
A=np.array([[-1,0,1],
           [3,0,-3],
           [1,0,-1]])
eig=np.linalg.eig(A)
print(eig)
eigvalues,eigvectors=eig[0],eig[1]

(array([ 0.,  0., -2.]), array([[ 0.        ,  0.70710678, -0.30151134],
       [ 1.        ,  0.        ,  0.90453403],
       [ 0.        ,  0.70710678,  0.30151134]]))


Compute $ A^{10} $  if A = 
$$
\left [
\begin{matrix}
-1 & 6  \\
1 & 0  \\
\end{matrix}
\right ]
$$

A has distinct eigenvalue: -3、2

In [1]:
import numpy as np
A=np.array([[-1,6],
            [1,0]])
np.linalg.eig(A)[0]

array([-3.,  2.])

According to Theorem 4.25: if A is an n$\times$n matrix with n distinct eigenvalues, then A is diagonalizable, there exists a diagonal matrix D, such that D=$ P^{-1} $$ A^{} $$ P^{} $ and A is similar to D.

According to theorem 4.22(e): if A~D, then A and D has the same eigenvalues, D has eigenvalues: -3、2

According to Theorem 4.15: the eigenvalues of triangle matrix are the entries on its main diagonal, diagonal matrix D is formed to be:
$$
\left [
\begin{matrix}
-3 & 0  \\
0 & 2  \\
\end{matrix}
\right ]
$$

In [2]:
D=np.array([[-3,0],
            [0,2]])

Given $ A^{} $$ P^{} $=$ P^{} $$ D^{} $, $ P\ne0 $:
\begin{equation}
A{\left [
\begin{matrix}
P_{1} & P_{2}  \\
\end{matrix}
\right ]}=\left [
\begin{matrix}
P_{1} & P_{2}  \\
\end{matrix}
\right ]\left [
\begin{matrix}
-3 & 0  \\
0 & 2  \\
\end{matrix}
\right ]
\end{equation}  
&nbsp;  
\begin{equation}
\left [
\begin{matrix}
AP_{1} & AP_{2}  \\
\end{matrix}
\right ]=\left [
\begin{matrix}
-3P_{1} & 2P_{2}  \\
\end{matrix}
\right ]
\end{equation}  
&nbsp;  
$$ AP_{1}=-3P_{1}, AP_{2}=2P_{2} $$  
&nbsp;  
the column vectors of P are eigenvectors of A whose corresponding eigenvalues are the diagonal entries of D in the same order.

In [3]:
import sympy as sp
sp.init_printing()

P=np.linalg.eig(A)[1]
sp.Matrix(P)

⎡-0.948683298050514  -0.894427190999916⎤
⎢                                      ⎥
⎣0.316227766016838   -0.447213595499958⎦

According to Theorem 4.22 (f): if A and B be n$\times$n matrices with A~B, then $ A^{m} $~$ B^{m} $, $ D^{10} $=$ P^{-1} $$ A^{10} $$ P^{} $

Given $ D^{10} $=$ P^{-1} $$ A^{10} $$ P^{} $, $ A^{10} $=$ P^{} $$ D^{10} $$ P^{-1} $

Given D is a diagonal matrix, $ D^{10} $ can also be a diagonal matrix, and all its entries in the principle diagonal are to the power of 10 

In [4]:
D_10=D**10
A_10= P @ D_10 @ np.linalg.inv(P)
A_10

array([[ 35839., -69630.],
       [-11605.,  24234.]])

$ A^{10} $ is formed to be :
$$
\left [
\begin{matrix}
35839 & -69630  \\
-11605 & 24234  \\
\end{matrix}
\right ]
$$